rllib/algorithms/impala/impala_torch_policy.py

```diff
@@ -172,6 +176,39 @@ def __init__(self):
     def optimizer(
         self,
     ) -> Union[List["torch.optim.Optimizer"], "torch.optim.Optimizer"]:
+
+        if self.config["_separate_vf_optimizer"]:
+            # Figure out, which parameters of the model belong to the value
+            # function (and which to the policy net).
+            dummy_batch = self._lazy_tensor_dict(
+                self._get_dummy_batch_from_view_requirements()
+            )
+            # Zero out all gradients (set to None)
+            for param in self.model.parameters():
+                param.grad = None
+            # Perform a dummy forward pass (through the policy net, which should be
+            # separated from the value function in this particular user setup).
+            out = self.model(dummy_batch)
+            # Perform a (dummy) backward pass to be able to see, which params have
+            # gradients and are therefore used for the policy computations (vs vf
+            # computations).
+            torch.sum(out[0]).backward()  # [0] -> Model returns out and state-outs.
+            # Collect policy vs value function params separately.
+            policy_params = []
+            value_params = []
+            for param in self.model.parameters():
+                if param.grad is None:
+                    value_params.append(param)
+                else:
+                    policy_params.append(param)
```

Comment:
I don't understand the need for this. Why can't you directly index to the model and ask it to give you `.value.parameters()` and `.policy.parameters()`? There should be a better way than treating `self.model` as a blackbox with only a knowledgt hat if I do forward pass on the model directly it will call the parameters that are used for `policy`. Also what if there are shared parameters between the value and policy components? This will lump them up into the policy's optimizer.  They won't get updated based on the loss from value function.

Reply:
Thanks for the explanation, I figured that might be the reason. We should be explicit about this in the comments